### PR TITLE
[core] Replace boost iostream with std::{i,s}stream

### LINKED ISSUE #4915: Remove boost iostreams dependency
jpeg_reader.cpp and png_reader.cpp use boost iostreams for simple buffer reads. Bringing in boost requires some warning-disabling boilerplate and even then causes a bunch of compile spew about deprecated headers.

It looks like the use of boost can be replaced with a simple buffer and current location; let's do so.

### PR DESCRIPTION
**Pull request by brunoabinader**
Fixes #4915.

:eyes: @jfirebaugh @springmeyer

### CODE REVIEW

**Comment by springmeyer** [comment]
👍

**Comment by jfirebaugh** [comment]
🙇

File: platform/default/png_reader.cpp [hunk 1]
```diff
@@ -54,8 +44,8 @@ struct png_struct_guard {
 };
 
 PremultipliedImage decodePNG(const uint8_t* data, size_t size) {
-    source_type source(reinterpret_cast<const char*>(data), size);
-    input_stream stream(source);
+    std::istringstream source(std::string(reinterpret_cast<const char*>(data), size));
```
Comment (artemp) [inline platform/default/png_reader.cpp]: ^ creates a new std::string and copies data!

File: platform/default/png_reader.cpp [hunk 2]
```diff
@@ -54,8 +44,8 @@ struct png_struct_guard {
...
+    std::istream stream(source.rdbuf());
```
Comment (artemp) [inline platform/default/png_reader.cpp]: hmm and again ^.

**Comment by artemp** [comment]
Better solution is to implement custom stream buffer : https://artofcode.wordpress.com/2010/12/12/deriving-from-stdstreambuf/
/cc @brunoabinader @jfirebaugh @springmeyer

**Comment by artemp** [comment]
@brunoabinader  This is what II did in mapnik https://github.com/mapnik/mapnik/commit/2e8c0d36c2237f2815d8004c1b96bad909056eb9
Note: for `tiff_reader` to work correctly stream buffer must implement 
`c++
pos_type pubseekoff( off_type off, ios_base::seekdir dir,
                     ios_base::openmode which = ios_base::in | ios_base::out );
```
http://en.cppreference.com/w/cpp/io/basic_streambuf/pubseekoff

/cc @springmeyer @jfirebaugh

**Comment by jfirebaugh** [comment]
@artemp I was wondering about that, thanks for the links.

**Comment by mb12** [comment]
@artemp and @jfirebaugh Is the concern here that duplicate copy of image data in std::string can cause performance issue? This may not be perceptible to the end user at all esp for Vector maps where this code will kick in only once when loading the sprite. #3048 is more likely to be a performance bottleneck arising due to copying data multiple times (But it has not been measured).

When we switched from Curl to OkHttp for network access, it introduced multiple copies of tile data as it moves across JNI boundary. (Issue #3048).

https://github.com/mapbox/mapbox-gl-native/issues/3048

**Comment by brunoabinader** [comment]
Thank you @artemp - I wonder if we could add `char_array_buffer` on its own header-only library so we could share code and avoid duplication? /cc @jfirebaugh

**Pull request merged**